### ISSUE TITLE
Blockstream Jade hangs on Processing when signing multi-sig transactions

### ISSUE THREAD

**Issue by shaun-widmer** [sparrowwallet/sparrow#1060]
Every part of the signing process works on the Jade device (verifying amount of BTC for sending, the fees, etc...) until the very last step in the signing process. It just says "Processing..." on the Jade device but never finished signing the transaction.

**My setup:**

- Using Sparrow 1.7.8.
- Attempting to sign with the USB cable connection
- Mac OS Big Sur 11.6 (Mac M1) - (I already installed the additional blockstream jade drivers from the blockstream website for my version of Mac)
- I have the latest version of firmware installed on the Jade wallet.
- I used blockstream green to initialize the jade and I set up the standard single sig wallet on the jade in Blockstream green. 
- See screenshot of multi-sig configuration below:

![Screen Shot 2023-08-09 at 5 03 02 PM](https://github.com/sparrowwallet/sparrow/assets/141878515/ab04dcba-18a9-4511-a98d-5275126fdfa5)

**Steps:**

1) I plug in the jade and then scan for the device in sparrow.
2) I unlock the jade with my passphrase and sparrow shows it as unlocked .
3) I click “sign” in sparrow and verify the transaction details on the jade and click the checkmark button to finalize the signing on the jade…

<img width="668" alt="Screen Shot 2023-08-09 at 5 26 15 PM" src="https://github.com/sparrowwallet/sparrow/assets/141878515/eb5f1b4e-8ecc-4c7a-88f5-fef5f537f5f0">

4) It just says “Processing…” on the jade… Then after a minute or 2, Sparrow shows the following error:

<img width="695" alt="Screen Shot 2023-08-09 at 5 26 31 PM" src="https://github.com/sparrowwallet/sparrow/assets/141878515/4f96abeb-366f-4763-8109-be6807e95189">

<img width="494" alt="Screen Shot 2023-08-09 at 5 13 46 PM" src="https://github.com/sparrowwallet/sparrow/assets/141878515/a53c8b03-dc49-4c67-ad49-ef014d8544b3">

I created a single sig wallet on Sparrow to test signing with the same jade device and it works flawlessly... but in the multi sig it doesn't work.

I purchased multiple jades and none of them work in a multi sign setup. Tried 10+ times.

**Comment by craigraw** [comment]
It appears that this is a Jade error, and not a Sparrow one. At a guess, if there are many inputs in this transaction, the Jade may have run out of memory while processing. I have successfully tested multisig signing on the Jade with both USB and QR, so it may be specific to this particular transaction.

@JamieDriver any thoughts?

**Comment by JamieDriver** [comment]
Yes I too have tested multisig regularly without issue - on linux at least.

However, there have been issues sending large amounts of data (eg. large tx) with the default MacOS drivers - see [this article](https://help.blockstream.com/hc/en-us/articles/900005443223-Fix-issues-connecting-Jade-via-USB#h_01H4PFV1R8CC9TCSBBMBTP5E6J).

FYI: To have got as far as seeing the outputs on the screen and confirming, the new spending tx has been successfully sent to Jade.  In the 'Processing' step Sparrow is sending each prior tx (ie. the tx that created the utxo being signed/spent) over to Jade.  It could be that one of those prior tx's is 'large'.

Another alternative maybe is to sign via QR ?

cc: @bitcoinhelp

**Comment by bitcoinhelp** [comment]
Also trying here and no issues, curious what happens if you try to spend a different utxo? (re: Jamie's comment - " It could be that one of those prior tx's is 'large'.")

Also any luck with QRs?

**Comment by craigraw** [comment]
@JamieDriver From my testing it appears that Jade does not require the PSBT to contain the `NON_WITNESS_UTXO` field when signing a segwit transaction ([as opposed to the Trezor which does](https://blog.trezor.io/details-of-firmware-updates-for-trezor-one-version-1-9-1-and-trezor-model-t-version-2-3-1-1eba8f60f2dd)).

Excluding this field would substantially reduce the PSBT size when signing segwit transactions, which might solve this issue. If you can confirm that the Jade does not take the same approach as the Trezor wrt the 2020 issue reported by Saleem Rashid (requiring the `NON_WITNESS_UTXO` field regardless of transaction type), I am happy to remove this field from PSBTs created for the Jade.

**Comment by JamieDriver** [comment]
Aha well, there's a rub!  

We do require the prior txn (ie. the NON_WITNESS_UTXO or equivalent) when using the original `sign_tx` api/protocol (which eg hwi uses) - yes indeed because of the issue you mention.  (Unless it's only a single segwit input, then it's optional.)

When we came to adding the PSBT signing via QRs, we found that [at least some of] the wallets we were testing with did not always include the `NON_WITNESS_UTXO` field in the PSBT QR, so we relaxed that check in that case.  (This is not ideal as it suggests that exploit is potentially open via PSBT/QR signing.)

But if you're using HWI or Jade's 'sign_tx' API directly, it will require the NON_WITNESS_UTXO if there are multiple inputs (or if the inputs are non-segwit).

ie. in your test above you may have hit the 'only one input, and it's segwit' case, in which case checking the prior tx is not necessary.

How do you communicate with Jade over USB, via hwilib ?

**Comment by craigraw** [comment]
Yes, using HWI. I've tested using two inputs and no `NON_WITNESS_UTXO` and indeed the signing does fail with `Failed to extract input_tx from parameters`. Thanks for clearing that up, I'll keep things as is. It looks like the solution in this case is to use QR. Fwiw I have always felt that the Trezor solution was pretty onerous for what is (in my opinion) an unlikely exploit.

**Comment by shaun-widmer** [comment]
Hey Guys, so I just checked and the QR scan option does in-fact work to sign the same multi sig transaction. (Thank you all very much for your help on this issue!)

It sure would be nice to use the USB option though because the QR scanning can be tough depending on what computer and webcam setup you have at the time of signing the transaction. 

Thoughts?

**Comment by craigraw** [comment]
I don't think there is anything further that can be done on the Sparrow side at this time. The issue you are running into is a hardware limitation. Due to a potential (possible although unlikely) exploit, the Jade requires the entirety of all transactions that created the outputs you are spending in the PSBT sent over USB for signing. If there are many multisig transactions making up these inputs, the size of the PSBT can rise quite substantially, and the Jade can run out of memory. Due to the unlikelihood of the exploit, and the impracticality of sending so much data over QR, this additional data is not sent over QR the this issue does not present itself. It should also be noted that Coldcard etc have found different and less onerous solutions to the potential exploit.

Closing this off here.

**Comment by JamieDriver** [comment]
> It should also be noted that Coldcard etc have found different and less onerous solutions to the potential exploit.

Really?  I'll have to have a dig!  ;-)

**Comment by JamieDriver** [comment]
I don't think Jade is running out of memory or hitting any hardware limitation - I suspect the issue is the default/shipped MacOS driver that we have seen issues with before - basically it seems to corrupt 'large' messages sent from host to Jade over the serial interface, as mentioned [above](https://github.com/sparrowwallet/sparrow/issues/1060#issuecomment-1673077810).  

We discuss it [here](https://help.blockstream.com/hc/en-us/articles/900005443223-Fix-issues-connecting-Jade-via-USB#h_01H4PFV1R8CC9TCSBBMBTP5E6J), and suggest an improved driver - if you are open to trying it out it'd be good to confirm if it resolves the issue.

Thanks.

**Comment by craigraw** [comment]
> I don't think Jade is running out of memory or hitting any hardware limitation - I suspect the issue is the default/shipped MacOS driver that we have seen issues with before

Ok - thanks for clarifying!

> Really? I'll have to have a dig! ;-)

Here you go: https://blog.coinkite.com/version-3.1.4-released/

**Comment by shaun-widmer** [comment]
Jamie

> I don't think Jade is running out of memory or hitting any hardware limitation - I suspect the issue is the default/shipped MacOS driver that we have seen issues with before - basically it seems to corrupt 'large' messages sent from host to Jade over the serial interface, as mentioned [above](https://github.com/sparrowwallet/sparrow/issues/1060#issuecomment-1673077810).
> 
> We discuss it [here](https://help.blockstream.com/hc/en-us/articles/900005443223-Fix-issues-connecting-Jade-via-USB#h_01H4PFV1R8CC9TCSBBMBTP5E6J), and suggest an improved driver - if you are open to trying it out it'd be good to confirm if it resolves the issue.
> 
> Thanks.

Hey @JamieDriver, thanks for the tip... I downloaded both drivers for Mac listed there on the blockstream page. I rebooted my system then attempted to sign a multi sig transaction with the jade. For some reason, when I plug in my jade now via USB to sign, it shows 2 Jade devices that i can sign with. Anyways, the first jade device in the list WORKS to sign. THe 2nd one does not. Thanks for the help guys!

<img width="509" alt="Screen Shot 2023-08-14 at 5 06 13 PM" src="https://github.com/sparrowwallet/sparrow/assets/141878515/50571945-359d-4f0c-a65b-651a4b0a3517">

**Comment by shaun-widmer** [comment]
Hey @JamieDriver, thanks for the tip... I downloaded both drivers for Mac listed there on the blockstream page. I rebooted my system then attempted to sign a multi sig transaction with the jade. For some reason, when I plug in my jade now via USB to sign, it shows 2 Jade devices that i can sign with. Anyways, the first jade device in the list WORKS to sign. THe 2nd one does not. Thanks for the help guys!

![fre2](https://github.com/sparrowwallet/sparrow/assets/141878515/3af12982-a01e-4f45-828d-61f26400aeee)
